[OpenMP] Use simple VLA implementation to replace uses of actual VLA

shiltian opened the pull request:
Use of VLA can cause compile warning that was introduced in D156565. This patch
implements a simple stack/heap-based VLA that can miminc the behavior of an
actual VLA and prevent the warning. By default the stack accomodates the
elements. If the number of emelements is greater than N, which by default is 8,
a heap buffer will be allocated and used to acccomodate the elements.

unterumarmung commented on openmp/runtime/src/kmp_utils.h:
```diff
@@ -0,0 +1,39 @@
+  SimpleVLA(std::size_t Size) noexcept : Data(new T[Size]) {}
+
+  ~SimpleVLA() noexcept {
+    if (Data)
+      delete[] Data;
```
```suggestion
    delete[] Data;
```
`delete nullptr` is valid expression in C++ and is no-op. Hence, the `if` statement is not needed.

unterumarmung commented on openmp/runtime/src/kmp_utils.h:
```diff
@@ -0,0 +1,39 @@
+public:
+  SimpleVLA(std::size_t Size) noexcept : Data(new T[Size]) {}
+
+  ~SimpleVLA() noexcept {
```
```suggestion
  ~SimpleVLA() {
```
Destructors are `noexcept` by default

unterumarmung commented on openmp/runtime/src/kmp_utils.h:
```diff
@@ -0,0 +1,39 @@
+  const T &operator[](std::size_t Idx) const { return Data[Idx]; }
+  T &operator[](std::size_t Idx) { return Data[Idx]; }
+
+  operator T *() { return Data; }
+  operator const T *() const { return Data; }
```
Should these functions be `noexcept` also?

unterumarmung commented on openmp/runtime/src/kmp_utils.h:
```diff
@@ -0,0 +1,39 @@
+  T *Data = nullptr;
+
+public:
+  SimpleVLA(std::size_t Size) noexcept : Data(new T[Size]) {}
```
```suggestion
  explicit SimpleVLA(std::size_t Size) noexcept : Data(new T[Size]) {}
```
I guess, it is not a desired behavior where an integer can implicitly create an array.

unterumarmung replied:
Also, I'd suggest to explicitly define copy/move constructors and the default one.
For example,
```cpp
  SimpleVLA() = delete;
  SimpleVLA(const SimpleVLA&) = delete;
  SimpleVLA(SimpleVLA&&) = default;
  SimpleVLA& operator=(const SimpleVLA&) = delete;
  SimpleVLA& operator=(SimpleVLA&&) = default;
```

unterumarmung reviewed (commented):
I'm not an OpenMP contributor/maintainer/developer, so feel free to dismiss my suggestions.

shiltian commented:
> I'm not an OpenMP contributor/maintainer/developer, so feel free to dismiss my suggestions.

Thanks for the comments. They are very helpful!

unterumarmung commented:
> > I'm not an OpenMP contributor/maintainer/developer, so feel free to dismiss my suggestions.
> 
> Thanks for the comments. They are very helpful!

I think you've missed the `explicit` suggestion, I think it is important..

unterumarmung commented on openmp/runtime/src/kmp_utils.h:
```diff
@@ -0,0 +1,42 @@
+
+  ~SimpleVLA() { delete[] Data; }
+
+  const T &operator[](std::size_t Idx) const noexcept { return Data[Idx]; }
+  T &operator[](std::size_t Idx) noexcept { return Data[Idx]; }
```
I'm not sure if these ones should be marked `noexcept`. Every array-like C++ data structure (vector, string and array) does not mark them as `noexcept` because there can be possibly an UB (where UB = can happen anything). Here, it may be ok, may be not, I don't know.

frederick-vs-ja replied:
> I'm not sure if these ones should be marked `noexcept`. Every array-like C++ data structure (vector, string and array) does not mark them as `noexcept` because there can be possibly an UB (where UB = can happen anything). Here, it may be ok, may be not, I don't know.

The purpose of not marking `noexcept` is to allow implementations to diagnose UB by throwing exceptions. This is not the case for user codes.

frederick-vs-ja commented on openmp/runtime/src/kmp_utils.h:
```diff
@@ -0,0 +1,42 @@
+public:
+  SimpleVLA() = delete;
+  SimpleVLA(const SimpleVLA &) = delete;
+  SimpleVLA(SimpleVLA &&) = default;
+  SimpleVLA &operator=(const SimpleVLA &) = delete;
+  SimpleVLA &operator=(SimpleVLA &&) = default;
```
Defaulted move functions are still problematic. Once we move constructs a new `SimpleVLA` from the old one, the old one will hold the data, and double-free will probably happen.

Can we simply reuse `unique_ptr<T[]>` and `make_unique`?

shiltian replied:
I'm not sure if the smart pointer will introduce the dependence on any form of C++ library since we can't do that.

frederick-vs-ja replied:
I suppose it's OK to use `unique_ptr`, given the current stategy is already using `operator new[]` and `operator delete[]` form the C++ standard library.

shiltian replied:
FWIW, we already use `operator new[]` in `libomp` but it doesn't look like it actually uses things from C++ standard library.

shiltian replied:
I removed the move functions because I don't think that is allowed by actual VLA.

jdoerfert reviewed (commented):
Can we have a very simple, non-growable, SmallVector-like API?
So provide a compile time bound for the size that could live on the stack, and if it's too large at runtime do a alloc/free.
The cost for this is basically nothing, but you avoid potential allocation cost for 95% of all cases (size <= 16 or 32).

shiltian committed:
[OpenMP] Use simple VLA implementation to replace uses of actual VLA

Use of VLA can cause compile warning that was introduced in D156565. This patch
implements a simple stack/heap-based VLA that can miminc the behavior of an
actual VLA and prevent the warning. By default the stack accomodates the
elements. If the number of emelements is greater than N, which by default is 8,
a heap buffer will be allocated and used to acccomodate the elements.

shiltian commented:
Gentle ping

jdoerfert reviewed (approved):
LG, thx
